src/Compilers/CSharp/Test/Symbol/Symbols/CovariantReturnTests.cs

```diff
@@ -4436,5 +4456,305 @@ void verify(CSharpCompilation compilation)
+}
+";
+            var comp = CreateCompilation(withCovariantReturns: withCovariantReturns, source);
+            comp.VerifyDiagnostics(
+                // (5,25): error CS0111: Type 'Base<T>' already defines a member called 'M' with the same parameter types
+                //     public virtual void M(int y) => throw null; // 2
+                Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M").WithArguments("M", "Base<T>").WithLocation(5, 25),
+                // (10,26): error CS0462: The inherited members 'Base<T>.M(int)' and 'Base<T>.M(int)' have the same signature in type 'Derived', so they cannot be overridden
+                //     public override void M(int z) => throw null; // 3
+                Diagnostic(ErrorCode.ERR_AmbigOverride, "M").WithArguments("Base<T>.M(int)", "Base<T>.M(int)", "Derived").WithLocation(10, 26)
+                );
+
+            var member = (SourceMethodSymbol)comp.GlobalNamespace.GetMember("Derived.M");
+            bool useMethodImpl = member.RequiresExplicitOverride(out bool shouldWarn);
+            Assert.False(shouldWarn);
+            Assert.Equal(withCovariantReturns, useMethodImpl);
+
+            verify(SourceView(comp, ""));
+            verify(CompilationReferenceView(comp, ""));
+            verify(RetargetingView(comp, ""));
+
+            void verify(CSharpCompilation compilation)
+            {
+                VerifyOverride(compilation,
+                    methodName: "Derived.M",
+                    overridingMemberDisplay: "void Derived.M(System.Int32 z)",
+                    overriddenMemberDisplay: "void Base<System.Int32>.M(System.Int32 x)",
+                    requiresMethodimpl: useMethodImpl);
+            }
+        }
+
+        [Theory]
+        [CombinatorialData]
+        public void DuplicateDeclarations_03(
+            bool withCovariantReturns
+            )
+        {
+            var source = @"
+class Container<T>
+{
+    class Base
+    {
+        public virtual void M(int x) => throw null; // 1
+        public virtual void M(int y) => throw null; // 2
+    }
+
+    class Derived : Base
+    {
+        public override void M(int z) => throw null; // 3
+    }
+}
+";
+            var comp = CreateCompilation(withCovariantReturns: withCovariantReturns, source);
+            comp.VerifyDiagnostics(
+                // (7,29): error CS0111: Type 'Container<T>.Base' already defines a member called 'M' with the same parameter types
+                //         public virtual void M(int y) => throw null; // 2
+                Diagnostic(ErrorCode.ERR_MemberAlreadyExists, "M").WithArguments("M", "Container<T>.Base").WithLocation(7, 29)
+                );
+
+            var member = (SourceMethodSymbol)comp.GlobalNamespace.GetMember("Container.Derived.M");
+            bool useMethodImpl = member.RequiresExplicitOverride(out bool shouldWarn);
+            Assert.False(shouldWarn);
+            Assert.Equal(withCovariantReturns, useMethodImpl);
```

Comment:
>useMethodImpl [](start = 47, length = 13)

This should always be false

Reply:
Incorrect. This is the correct value by design. Specifically, when compiling targeting a platform that contains a fix for [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119), the method `RequiresExplicitOverride` returns true if the type containing the overridden method contains runtime-ambiguous methods.

In any case, this PR does not change the implementation of this value. It only tests it to confirm it is as design and expected. Which it is.

---
In reply to: [452976349](https://github.com/dotnet/roslyn/pull/45849#discussion_r452976349) [](ancestors = 452976349)

Reply:
> This is the correct value by design.

You keep referring to some design. Could you please provide concrete information what design decision you are referring to, who and when made it? Perhaps exact quote of the decision might help.

> Specifically, when compiling targeting a platform that contains a fix for [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119), the method RequiresExplicitOverride returns true if the type containing the overridden method contains runtime-ambiguous methods.

I do not think I agree with that. I believe the bug in the runtime is completely orthogonal to the scenarios that this PR is focusing. The reason is - it makes absolutely no difference if runtime has the fix. Emitting a method impl for these cases didn't help before and doing this doesn't help with the runtime fix. In fact we do not emit any code for these scenarios. Unnecessarily adding conditional behavior to the compiler for these scenarios is a big negative effect from the behavior proposed by this PR. This increases complexity of the compiler without providing any benefits, in my opinion. 

---
In reply to: [453786435](https://github.com/dotnet/roslyn/pull/45849#discussion_r453786435) [](ancestors = 453786435,452976349)

Reply:
The code reflects my design intent. You can read a detailed description of my decisions by reading the code. For this scenario, I would quote "when compiling targeting a platform that contains a fix for [https://github.com/dotnet/runtime/issues/38119](https://github.com/dotnet/runtime/issues/38119), the method RequiresExplicitOverride returns true if the type containing the overridden method contains runtime-ambiguous methods". That is part of the design intent here. I agree that unnecessarily adding conditional behavior to the compiler as you suggest is a big negative and I do not intend to do so.

If you believe this PR does not suppress the cascaded diagnostics as intended then please say so.

---
In reply to: [453807985](https://github.com/dotnet/roslyn/pull/45849#discussion_r453807985) [](ancestors = 453807985,453786435,452976349)